custom_components/amplipi/media_player.py

```diff
@@ -1235,3 +1220,273 @@ async def async_set_volume_level(self, volume):
+        self._current_source = None
+        self._current_zones: List[Zone] = []
+        self._current_groups: List[Group] = []
+        self._sources = sources
+        self._domain = namespace
+
+        self._id = stream.id
+        self._name = stream.name
+        self._unique_id = stream.unique_id
+        self.entity_id = stream.entity_id
+        # not a real device class, but allows us to match streams and only streams with the start_streaming blueprint's streams dropdown
+        self._attr_device_class = "stream"
+        
+        self._image_base_path = image_base_path
+        self._vendor = vendor
+        self._version = version
+        self._client = client
+        self._attr_source_list = [
+            'None',
+            'Any',
+            'Source 1',
+            'Source 2',
+            'Source 3',
+            'Source 4',
+        ] if self._stream.type != "rca" else ['None', f'Source {self._stream.id - 995}']
+        self._available = False
+        self._extra_attributes = []
+        self._is_off: bool = True
+
+    async def _update_zones(self, update: ZoneUpdate):
+        if self._current_source is not None:
+            multi_update = MultiZoneUpdate(
+                groups=[g.id for g in self._current_groups],
+                zones=[z.id for z in self._current_zones],
+                update=update
+            )
+            await self._client.set_zones(multi_update)
+            
+
+    async def async_toggle(self):
+        if not self._is_off or self._current_source is not None:
+            await self.async_turn_off()
+        else:
+            await self.async_turn_on()
+
+    async def async_turn_on(self):
+        if await self.find_source(): # Autoconnect stream if there is space
+            await self.async_connect_stream_to_source(self._stream)
+        elif self._stream.type == "rca": # Autoconnect stream if it can only ever connect to one source, regardless of space
+            goal_source = next((s for s in self._sources if s.id == self._id - 996), None)
+            if goal_source:
+                await self.async_connect_stream_to_source(self._stream, goal_source)
+
+        self._is_off = False
+
+    async def async_turn_off(self):
+        try:
+            if self._current_source is not None:
+                _LOGGER.info(f"Disconnecting stream from source {self._current_source.name}")
+                await self._update_zones(
+                    ZoneUpdate(
+                        source_id=-1,
+                    )
+                )
+                await self._client.set_source(
+                    self._current_source.id,
+                    SourceUpdate(
+                        input='None'
+                    )
+                )
+        except AttributeError:
+            # There can be a race condition where self._current_source is not None and then becomes None by the time self._current_source.name or self._current_source.id is read
```

Comment:
Reminder that race conditions like this may be reduced by routing all of the api requests thought the coordinator so that responses update the state

Reply:
Yup. Just wanted to keep this on your radar..